File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.py
# 오픈채팅방 문제 해결 함수
def solution(record):
    user_dict = {}  # 사용자 ID와 최종 닉네임을 매핑할 딕셔너리
    actions = []  # 최종 메시지를 저장할 리스트
    
    # 첫 번째 순회: 사용자 ID와 최종 닉네임 매핑
    for r in record:
        parts = r.split(' ')  # 각 레코드를 공백으로 분리
        
        # Enter나 Change 액션의 경우 닉네임 업데이트
        if parts[0] in ["Enter", "Change"]:
            nickname = parts[2]  # 세 번째 요소는 닉네임
            # 사용자 ID와 닉네임 매핑 업데이트
            user_dict[parts[1]] = nickname  
    
    # 두 번째 순회: 최종 메시지 생성
    for r in record:
        parts = r.split(' ')  # 각 레코드를 다시 공백으로 분리
        
        # Enter 액션: 입장 메시지 생성
        if parts[0] == "Enter":
            actions.append(f"{user_dict[parts[1]]}님이 들어왔습니다.")
        # Leave 액션: 퇴장 메시지 생성
        elif parts[0] == "Leave":
            actions.append(f"{user_dict[parts[1]]}님이 나갔습니다.")
    
    return actions  # 최종 메시지 리스트 반환

File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.java
import java.util.HashMap;
import java.util.ArrayList;

class Solution {
    // solution 메서드: 채팅방 로그를 처리하여 최종 메시지를 반환
    public String[] solution(String[] record){
        // Enter/Leave 메시지를 저장할 해시맵 생성
        HashMap<String, String> msg = new HashMap<>();
        // Enter 와 Leave 명령에 대한 메시지 매핑
        msg.put("Enter", "님이 들어왔습니다.");
        msg.put("Leave", "님이 나갔습니다.");

        // uid 와 닉네임을 저장할 해시맵 생성
        HashMap<String, String> uid = new HashMap<>();

        // record 배열의 각 요소를 순회하며 닉네임을 갱신
        for (String s : record) {
            // 각 명령어를 공백 기준으로 분리
            String[] cmd = s.split(" ");
            // Enter 또는 Change 명령어에 대해 닉네임 갱신
            // Enter 랑 Change 가 있으면 cmd 의 length 가 3이다. if 문 사용
            if (cmd.length == 3) {
                uid.put(cmd[1], cmd[2]);
            }
        }

        // 답을 저장할 answer 리스트 생성
        ArrayList<String> answer = new ArrayList<>();

        // record 배열을 다시 순회하며 메시지를 처리
        for (String s:record){
            // 각 명령어를 공백 기준으로 분리
            String[] cmd = s.split(" ");
            // Enter 또는 Leave 명령어인 경우 메시지 생성
            if(msg.containsKey(cmd[0])){
                // 유저 ID에 해당하는 닉네임과 메시지를 리스트에 추가
                answer.add(uid.get(cmd[1]) + msg.get(cmd[0]));
            }
        }

        // 최종 메시지를 String 배열로 변환하여 반환
        return answer.toArray(new String[0]);
    }
}